### ISSUE TITLE
Squiz.Commenting.FunctionComment.TypeHintMissing produces false positives

### ISSUE THREAD

**Issue by mindplay-dk** [squizlabs/PHP_CodeSniffer#1159]
As mentioned in #1153, it appears that `Squiz.Commenting.FunctionComment.TypeHintMissing` is producing false positives, and/or this particular inspection is actually covering two different types of errors.

[Here is an example file](https://gist.github.com/mindplay-dk/be76016ba4ff328521ef7761470a8585),for which I received the following inspections:

```
<file name="C:\workspace\test\score\src\KodusClient.php">
 <error line="71" column="12" severity="error" message="Type hint &quot;array&quot; missing for $headers" source="Squiz.Commenting.FunctionComment.TypeHintMissing"/>
 <error line="83" column="12" severity="error" message="Type hint &quot;array&quot; missing for $headers" source="Squiz.Commenting.FunctionComment.TypeHintMissing"/>
 <error line="95" column="12" severity="error" message="Type hint &quot;array&quot; missing for $headers" source="Squiz.Commenting.FunctionComment.TypeHintMissing"/>
 <error line="107" column="12" severity="error" message="Type hint &quot;array&quot; missing for $headers" source="Squiz.Commenting.FunctionComment.TypeHintMissing"/>
</file>
```

All of these functions appear to be correctly type-hinted.

It appears this inspection may be catching two different cases: one where the type-hint is missing from the php-doc block, and another where a static type-hint is missing from the actual parameter declaration in the code?

These are two very different error-conditions.

For various different reasons, it's not always possible to type-hint correctly with static type-hint - less so in PHP 7, where we can type-hint scalar arguments, but there are still examples where you cannot type-hint statically for some reason, such as having to use an upper or lower bound on an argument you inherited from a generic interface.

Am I correct that these two types of error are being reported as the same error?

To be clear: all I want is an inspection for `@param` doc-blocks with no type-hint, such as `@param $foo` - for some reason Php Storm auto-generates empty (invalid?) `@param` tags, and our devs frequently forget to correct them... As for static type-hints on the other hand, as explained, we can't really enforce anything, as there would be literally hundreds of cases where devs would have to suppress these inspections with a tag, which isn't practical.

On a related note, it seems you can no longer override the PHP version in version 3.0? Previously, I was able to set a config option that would override the PHP version number to a lower version. This was useful, since building/testing on a local PHP 7 installation doesn't mean that's the lowest version I'm targeting; being able to specify the minimum target PHP version, is how I perceived this feature. It was useful in relation to the scalar type-hint issue cited above, since that one is conditioned on PHP version being >= 7.

(was it removed, or did it's name change, or?)

**Comment by gsherwood** [comment]
> It appears this inspection may be catching two different cases: one where the type-hint is missing from the php-doc block, and another where a static type-hint is missing from the actual parameter declaration in the code?

Yes, that is exactly how the sniff has been designed. If you specify in the docblock that the param can only be an array, it wants the type hint in the function declaration. If you type hint the param in the declaration, it wants to see that in the docblock. The job of this sniff error message is to keep them in sync.

**Comment by gsherwood** [comment]
> On a related note, it seems you can no longer override the PHP version in version 3.0?

I assume you mean this change, which was released in 2.7.0 a couple of weeks ago:
```
- A few sniffs that produce errors based on the current PHP version can now be told to run using a specific PHP version
    - Set the php_version config var using --config-set, --runtime-set, or in a ruleset to specify a specific PHP version
    - The format of the PHP version is the same as the PHP_VERSION_ID constant (e.g., 50403 for version 5.4.3)
    - Supported sniffs are Generic.PHP.DisallowAlternativePHPTags, PSR1.Classes.ClassDeclaration, Squiz.Commenting.FunctionComment
```
That `php_version` config option is also available in 3.0, for the same sniffs listed in the changelog above.
